src/qibo/tensorflow/custom_operators/cc/kernels/measurements.cc

```diff
@@ -0,0 +1,99 @@
+#if GOOGLE_CUDA
+#define EIGEN_USE_GPU
+#endif  // GOOGLE_CUDA
+
+#include "measurements.h"
+#include "tensorflow/core/framework/op_kernel.h"
+
+namespace tensorflow {
+
+typedef Eigen::ThreadPoolDevice CPUDevice;
+typedef Eigen::GpuDevice GPUDevice;
+
+namespace functor {
+
+// CPU specialization
+template <typename Tint, typename Tfloat>
+struct MeasureFrequenciesFunctor<CPUDevice, Tint, Tfloat> {
+  void operator()(const CPUDevice &d, Tint* frequencies, const Tfloat* cumprobs,
+                  int64 nshots, int nqubits)
+  {
+    int64 nstates = 1 << nqubits;
+    #pragma omp parallel shared(cumprobs)
+    {
+        std::map<int64, int64> frequencies_private;
+        #pragma omp for
+        for (auto i = 0; i < nshots; i++) {
+          Tfloat random_number = ((Tfloat) rand() / RAND_MAX);
+          for (auto j = 0; j < nstates; j++) {
+            if (random_number < cumprobs[j]) {
+                if (frequencies_private.find(j) == frequencies_private.end()) {
```

Comment:
I think, this is the killer check for the parallel evaluation.

Reply:
In principle (and if I am correct) the `map`/`unordered_map` should implement a hash table so lookups should be O(1). Alternatively one can just use a simple vector for `frequencies_private` and then add it to the total `frequencies` vector, I just thought the map approach would be more memory friendly for more qubits (because `frequencies` and `frequencies_private` would have size 2^nqubits).

What is more interesting though is that even the non-OMP version which uses a simple loop:
```C++
for (auto i = 0; i < nshots; i++) {
    Tfloat random_number = ((Tfloat) rand() / RAND_MAX);
    for (auto j = 0; j < nstates; j++) {
        if (random_number < cumprobs[j]) {
            frequencies_private[j]++;
            break;
        }
    }
}
```
is slower than #325, which is also single thread but much more memory intensive.